### PR TITLE
Add BlockMask varlen fields and create_varlen_block_mask function

### PR DESCRIPTION
**Pull request by v0i0**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #173020
* #173019
* #173018
* __->__ #173016

Add support for variable-length sequences in flex_attention by extending
the BlockMask data structure with offset and limit fields:
- kv_offsets, kv_limits: per-block offsets and limits for key/value
- q_offsets, q_limits: per-block offsets and limits for query

Add create_varlen_block_mask() function that constructs BlockMask objects
for variable-length document packing. This enables batch-invariant numerics
by aligning sequences to block boundaries.

Add _setup_varlen_offsets() helper to the inductor kernel for handling
the new BlockMask fields, and update block_mask tuple unpacking to include
the new fields.

Add unit tests for create_varlen_block_mask() that verify the block mask
structure without invoking actual kernels.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos @Chillee @drisspg @yanboliang @BoyuanFeng

### CODE REVIEW

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -78,6 +78,34 @@ def _permute_strides(out: torch.Tensor, query_strides: tuple[int, ...]) -> torch
...
+    block_mask: tuple,
+) -> tuple[int, int, Tensor, Tensor, Tensor, Tensor, int, int]:
+    """Extract varlen parameters from block_mask tuple.
+
+    Returns:
+        (logical_q_len, logical_kv_len, q_offsets, kv_offsets,
+         q_limits, kv_limits, Q_BLOCK_SIZE, KV_BLOCK_SIZE)
+    """
+    return (
+        block_mask[0],  # logical_q_len
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:98]: is there any fancy named-tuple shenanigans we could do here to make this a little less sketchy?

File: torch/_inductor/kernel/flex/flex_attention.py [hunk 2]
```diff
@@ -68,6 +68,81 @@ def _sanitize_kernel_options_for_triton(
...
+    else:
+        # For varlen, use the logical sequence lengths from the block_mask
+        # These are set by create_varlen_block_mask and represent the padded iteration space
+        logical_seq_len_q = block_mask_q_length
+        logical_seq_len_kv = block_mask_kv_length
+
+    # Create scalar tensor inputs for logical lengths (only used when HAS_OFFSETS=true)
+    # These are loaded at runtime to avoid constexpr recompilation for each unique length
+    if has_offsets:
+        logical_q_len_buf = V.graph.add_tensor_constant(
```
Comment (drisspg) [inline torch/_inductor/kernel/flex/flex_attention.py:115]: maybe I should look at the the upper stack, but these need to be device side tensors and not launch params, for cuda graphs? is there other reasons?

File: torch/nn/attention/flex_attention.py [hunk 3]
```diff
@@ -499,6 +500,11 @@ class BlockMask:
     BLOCK_SIZE: tuple[int, int]
     mask_mod: _mask_mod_signature
 
+    kv_offsets: Tensor | None
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:503]: can you update the doc block

File: torch/nn/attention/flex_attention.py [hunk 4]
```diff
@@ -1215,6 +1263,261 @@ def causal_mask(b, h, q_idx, kv_idx):
...
+    kv_seq_lens = kv_seq_lens.to(device=device)
+    num_docs = q_seq_lens.shape[0]
+
+    if kv_seq_lens.shape[0] != num_docs:
+        raise ValueError(
+            f"q_seq_lens and kv_seq_lens must have the same number of documents. "
+            f"Got {num_docs} and {kv_seq_lens.shape[0]}"
+        )
+
+    if (q_seq_lens <= 0).any() or (kv_seq_lens <= 0).any():
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:1355]: sync?

File: torch/nn/attention/flex_attention.py [hunk 5]
```diff
@@ -1215,6 +1263,261 @@ def causal_mask(b, h, q_idx, kv_idx):
...
+    # Physical offsets: cumsum of actual lengths (used for memory access)
+    q_physical_doc_offsets = torch.zeros(num_docs + 1, device=device, dtype=torch.int32)
+    kv_physical_doc_offsets = torch.zeros(
+        num_docs + 1, device=device, dtype=torch.int32
+    )
+    q_physical_doc_offsets[1:] = torch.cumsum(q_seq_lens, dim=0)
+    kv_physical_doc_offsets[1:] = torch.cumsum(kv_seq_lens, dim=0)
+
+    # Total implied sequence lengths and block counts (from precomputed cumsums)
+    total_q_blocks = int(q_cumsum_blocks[-1].item())
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:1383]: hmm I wonder if we should think of a way to provide extra info that can avoid syncs, since this will typically be called near the dataloader and some/all that info may live on cpu already
